aspnetcore/fundamentals/servers/kestrel.md

```diff
@@ -130,6 +132,21 @@ The following examples use the <xref:Microsoft.AspNetCore.Server.Kestrel.Core> n
 using Microsoft.AspNetCore.Server.Kestrel.Core;
 ```
 
+Kestrel options, which are configured in C# code in the following examples, can also be set using a [configuration provider](xref:fundamentals/configuration/index). For example, the File Configuration Provider can load Kestrel configuration from a file (*appsettings.json* or *appsettings.{Environment}.json*) when `CreateDefaultBuilder` is called:
```

Comment:
You probably want to add that you need to do the following to make this work. Pining @Tratcher.

```csharp
.ConfigureServices((context, serviceCollection) =>
{
    serviceCollection.Configure<KestrelServerOptions>(context.Configuration.GetSection("Kestrel"));
})
```

Reply:
Have you actually written the code to verify that these lines of code are need at all?

Reply:
Absolutely.

Can you provide a sample project without this code or something very similar that reads the AddServerHeader from appsettings.json? My understanding is that some option types are populated from config by default, but KestrelServerOptions is not one of those types.

Reply:
Sorry, I might have misunderstood or overlooked the `KestrelServerOptions`. I use the following in appsettings.json and never need to add that lines of code in program.cs:
```
    "Kestrel": {
        "EndpointDefaults": {
            "Protocols": ["Http2"]
        },
        "Endpoints": {
            "HttpsDefaultCert": {
                "Url": "https://*:5000",
                "Protocols": ["Http2"]
            }
        },
        "Certificates": {
            "Default": {
                "Path": "/tmp/localhost.pfx",
                "Password": "4xLabs.com"
            }
        }
    },
```
My understanding is that if *everything* is configurable from appsettings.json, there is no need to add c# code unless there is something to further customize or add or not possible to add in appsettings.json. Does `KestrelServerOptions` map to the `Kestrel` in that appsettings.json snippet I show or it is something different?

Reply:
Kestrel has [special logic](https://github.com/aspnet/AspNetCore/blob/ff8363a638401c5602910553356bd7f98b8d2f98/src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs#L216) to read endpoint and certificate options from config. It's not a direct mapping.

You can do a direct mapping to from config to KestrelServerOptions by calling `Configure<KestrelServerOptions>()` as demonstrated in my original comment.

Reply:
Can you provide example config which needs the `Configure<KestrelServerOptions>()`? I didn't have that line of C# code but the application still picks up the settings from appsetting.json alone.

Reply:
@halter73 I see ... so ur saying that relying on `CreateDefaultBuilder` > `ConfigureAppConfiguration` > `AddJsonFile` doesn't work because the Kestrel config doesn't **_look_** :eyes: for config unless it's explicitly told via `Configure` to grab and use the *Kestrel* section to do so. Gotcha. It seems a little surprising tho that it doesn't work OOB because ...

* If the dev has Kestrel options config in the files (or via env vars ... or via any config provider really), then they intend it to flow into Kestrel config auto-magically. They have to take this extra step. Why not just take whatever they supply?
* The *special logic* creates an inconsistency that has to be addressed in an already complex system (Kestrel server) and docs.

An additional point ... On the next commit, I'm going to make it say that `GetSection` will pull in configuration from any config provider.

Reply:
> @halter73 I see ... so ur saying that relying on `CreateDefaultBuilder` > `ConfigureAppConfiguration` > `AddJsonFile` doesn't work because the Kestrel config doesn't **_look_** 👀 for config unless it's explicitly told via `Configure` to grab and use the _Kestrel_ section to do so. Gotcha. It seems a little surprising tho that it doesn't work OOB because ...

No, there are two different things going on here:
- Options config: Standard config to options binding for setting simple properties like KestrelServerOptions.AddServerHeader. This is not enabled by default, you have to call `serviceCollection.Configure<KestrelServerOptions>(context.Configuration.GetSection("Kestrel"));` the same way you would for any other options config binding.
- Endpoint config: This is custom config binding implemented in Kestrel and enabled by the default Host. This maps complex configs like endpoint bindings as covered in this doc.

Reply:
I've almost got it updated. Let's look at the next commit and see where it goes right/wrong. Should have it in like five minutes. 🏃